### ISSUE TITLE
💅 `noUndeclaredDependencies` special case

### ISSUE THREAD

**Issue by mrazauskas** [biomejs/biome#4711]
### Environment information

```bash
CLI:
  Version:                      1.9.4
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_LOG_PATH:               unset
  BIOME_LOG_PREFIX_NAME:        unset
  BIOME_CONFIG_PATH:            unset
  NO_COLOR:                     unset
  TERM:                         "xterm-256color"
  JS_RUNTIME_VERSION:           "v22.11.0"
  JS_RUNTIME_NAME:              "node"
  NODE_PACKAGE_MANAGER:         "npm/10.9.0"

Biome Configuration:
  Status:                       Loaded successfully
  Formatter disabled:           false
  Linter disabled:              false
  Organize imports disabled:    false
  VCS disabled:                 true

Linter:
  JavaScript enabled:           true
  JSON enabled:                 true
  CSS enabled:                  true
  GraphQL enabled:              false
  Recommended:                  true
  All:                          false
  Enabled rules:
  performance/noDelete
  suspicious/noCatchAssign
  suspicious/noUnsafeNegation
  complexity/noForEach
  correctness/noUnusedImports
  complexity/noUselessStringConcat
  complexity/noMultipleSpacesInRegularExpressionLiterals
  a11y/useValidLang
  complexity/noUselessEmptyExport
  suspicious/useNamespaceKeyword
  suspicious/useValidTypeof
  style/useExplicitLengthCheck
  correctness/noConstantCondition
  a11y/useAriaActivedescendantWithTabindex
  a11y/useValidAriaRole
  style/useDefaultParameterLast
  complexity/noEmptyTypeParameters
  correctness/noConstructorReturn
  style/useSelfClosingElements
  suspicious/noAssignInExpressions
  suspicious/noDuplicateParameters
  suspicious/noDuplicateSelectorsKeyframeBlock
  correctness/noUnknownProperty
  style/useTemplate
  correctness/noUnusedLabels
  complexity/noUselessTernary
  suspicious/noExplicitAny
  correctness/noUnreachableSuper
  suspicious/noCompareNegZero
  correctness/noSwitchDeclarations
  a11y/noAutofocus
  correctness/noUnsafeOptionalChaining
  correctness/noConstAssign
  suspicious/noMisplacedAssertion
  suspicious/noControlCharactersInRegex
  complexity/noUselessTypeConstraint
  style/noVar
  suspicious/noDoubleEquals
  suspicious/noRedundantUseStrict
  style/useLiteralEnumMembers
  suspicious/noGlobalIsNan
  suspicious/noEmptyInterface
  suspicious/noSkippedTests
  style/useImportType
  suspicious/noMisleadingCharacterClass
  correctness/noPrecisionLoss
  a11y/noLabelWithoutControl
  suspicious/noRedeclare
  correctness/noStringCaseMismatch
  correctness/noSetterReturn
  correctness/noInvalidConstructorSuper
  suspicious/noImplicitAnyLet
  a11y/useKeyWithClickEvents
  suspicious/noDuplicateObjectKeys
  complexity/noUselessThisAlias
  correctness/noUnreachable
  suspicious/noFallthroughSwitchClause
  suspicious/noUnsafeDeclarationMerging
  complexity/noThisInStatic
  complexity/useOptionalChain
  correctness/noInnerDeclarations
  suspicious/noDuplicateCase
  a11y/useValidAnchor
  complexity/useRegexLiterals
  correctness/noSelfAssign
  correctness/noInvalidBuiltinInstantiation
  style/noUselessElse
  style/useShorthandFunctionType
  suspicious/noShadowRestrictedNames
  correctness/noInvalidDirectionInLinearGradient
  style/useThrowNewError
  suspicious/noImportantInKeyframe
  a11y/useMediaCaption
  complexity/noUselessLabel
  complexity/noUselessCatch
  correctness/noUnsafeFinally
  a11y/useAriaPropsForRole
  style/useCollapsedElseIf
  correctness/noNonoctalDecimalEscape
  style/useEnumInitializers
  a11y/useHtmlLang
  suspicious/noDuplicateTestHooks
  correctness/noUndeclaredDependencies
  suspicious/noEvolvingTypes
  style/useWhile
  complexity/noUselessUndefinedInitialization
  complexity/useArrowFunction
  style/noInferrableTypes
  a11y/noNoninteractiveTabindex
  complexity/useSimpleNumberKeys
  correctness/useYield
  a11y/noInteractiveElementToNoninteractiveRole
  style/useNumericLiterals
  correctness/noUnnecessaryContinue
  suspicious/noApproximativeNumericConstant
  suspicious/noImportAssign
  suspicious/noLabelVar
  correctness/noGlobalObjectCalls
  style/noParameterProperties
  suspicious/useDefaultSwitchClauseLast
  correctness/noUnknownUnit
  correctness/useImportExtensions
  a11y/useAltText
  correctness/noEmptyCharacterClassInRegex
  a11y/useIframeTitle
  suspicious/noPrototypeBuiltins
  a11y/noSvgWithoutTitle
  correctness/noVoidElementsWithChildren
  style/useAsConstAssertion
  suspicious/noDebugger
  style/useExportType
  complexity/noUselessLoneBlockStatements
  style/noArguments
  a11y/useValidAriaValues
  suspicious/noGlobalAssign
  suspicious/noCommentText
  a11y/useFocusableInteractive
  correctness/noUnmatchableAnbSelector
  suspicious/noMisleadingInstantiator
  suspicious/noDuplicateJsxProps
  suspicious/noThenProperty
  a11y/noPositiveTabindex
  correctness/noEmptyPattern
  complexity/noExcessiveNestedTestSuites
  security/noDangerouslySetInnerHtmlWithChildren
  a11y/useKeyWithMouseEvents
  style/useNamingConvention
  suspicious/noExtraNonNullAssertion
  correctness/noRenderReturnValue
  correctness/useExhaustiveDependencies
  security/noGlobalEval
  style/noNonNullAssertion
  a11y/noRedundantRoles
  complexity/useFlatMap
  correctness/useIsNan
  correctness/noUnusedVariables
  style/useConst
  style/noYodaExpression
  suspicious/noGlobalIsFinite
  style/useConsistentBuiltinInstantiation
  suspicious/noAsyncPromiseExecutor
  suspicious/noDuplicateFontNames
  suspicious/noSelfCompare
  security/noDangerouslySetInnerHtml
  style/useNodejsImportProtocol
  a11y/noDistractingElements
  style/useConsistentArrayType
  complexity/noWith
  suspicious/noArrayIndexKey
  style/useForOf
  suspicious/noDuplicateClassMembers
  complexity/noExtraBooleanCast
  performance/noAccumulatingSpread
  a11y/useValidAriaProps
  a11y/noRedundantAlt
  complexity/useDateNow
  correctness/noChildrenProp
  correctness/noInvalidPositionAtImportRule
  correctness/noUnknownFunction
  suspicious/noConfusingLabels
  suspicious/useGetterReturn
  suspicious/noConsole
  suspicious/noConfusingVoidType
  suspicious/noFocusedTests
  a11y/useButtonType
  a11y/useSemanticElements
  suspicious/noShorthandPropertyOverrides
  a11y/noAriaUnsupportedElements
  correctness/noInvalidGridAreas
  correctness/noFlatMapIdentity
  style/useShorthandAssign
  suspicious/useErrorMessage
  suspicious/noSuspiciousSemicolonInJsx
  a11y/noBlankTarget
  a11y/useHeadingContent
  suspicious/noSparseArray
  correctness/useValidForDirection
  correctness/noUnusedPrivateClassMembers
  correctness/noVoidTypeReturn
  correctness/noInvalidUseBeforeDeclaration
  a11y/noAriaHiddenOnFocusable
  a11y/useGenericFontNames
  correctness/noUnknownMediaFeatureName
  a11y/useAnchorContent
  complexity/noUselessRename
  style/useBlockStatements
  style/useNumberNamespace
  complexity/noUselessConstructor
  a11y/noAccessKey
  correctness/noUnusedFunctionParameters
  style/noUnusedTemplateLiteral
  complexity/noUselessSwitchCase
  style/useExponentiationOperator
  correctness/noUndeclaredVariables
  style/noNegationElse
  suspicious/useAwait
  style/useSingleVarDeclarator
  suspicious/noExportsInTest
  suspicious/useNumberToFixedDigitsArgument
  a11y/noNoninteractiveElementToInteractiveRole
  style/noCommaOperator
  suspicious/noDuplicateAtImportRules
  suspicious/useIsArray
  a11y/noHeaderScope
  complexity/noUselessFragments
  suspicious/noMisrefactoredShorthandAssign
  suspicious/noEmptyBlock
  suspicious/noClassAssign
  style/noImplicitBoolean
  suspicious/noFunctionAssign

Workspace:
  Open Documents:               0
```

### Rule name

`noUndeclaredDependencies`

### Playground link

https://biomejs.dev/playground/?lineWidth=120&indentStyle=space&lintRules=all&code=aQBtAHAAbwByAHQAIAB2AHMAYwBvAGQAZQAgAD0AIAByAGUAcQB1AGkAcgBlACgAIgB2AHMAYwBvAGQAZQAiACkAOwAKAAoAZQB4AHAAbwByAHQAIABmAHUAbgBjAHQAaQBvAG4AIABhAGMAdABpAHYAYQB0AGUAKABjAG8AbgB0AGUAeAB0ADoAIAB2AHMAYwBvAGQAZQAuAEUAeAB0AGUAbgBzAGkAbwBuAEMAbwBuAHQAZQB4AHQAKQA6ACAAdgBvAGkAZAAgAHsACgAgACAAYwBvAG4AcwB0ACAAZABpAHMAcABvAHMAYQBiAGwAZQAgAD0AIAB2AHMAYwBvAGQAZQAuAGMAbwBtAG0AYQBuAGQAcwAuAHIAZQBnAGkAcwB0AGUAcgBDAG8AbQBtAGEAbgBkACgAIgBoAGUAbABsAG8AdwBvAHIAbABkAC4AaABlAGwAbABvAFcAbwByAGwAZAAiACwAIAAoACkAIAA9AD4AIAB7AAoAIAAgACAAIAB2AHMAYwBvAGQAZQAuAHcAaQBuAGQAbwB3AC4AcwBoAG8AdwBJAG4AZgBvAHIAbQBhAHQAaQBvAG4ATQBlAHMAcwBhAGcAZQAoACIASABlAGwAbABvACAAVwBvAHIAbABkACEAIgApADsACgAgACAAfQApADsACgAKACAAIABjAG8AbgB0AGUAeAB0AC4AcwB1AGIAcwBjAHIAaQBwAHQAaQBvAG4AcwAuAHAAdQBzAGgAKABkAGkAcwBwAG8AcwBhAGIAbABlACkAOwAKAH0ACgA%3D

### Expected result

VS Code extensions must import APIs from `"vscode"` package. This is just a place holder, such package is not even published to `npm` registry.

Might be worth to treat `"vscode"` as a special case. I think it should not be treated as missing dependency.

--- 

By the way, it would nice to support the following syntax:

```ts
import vscode = require("vscode");
```

This way `tsc` emits cleaner code:

```js
const vscode = require("vscode");
```

### Code of Conduct

- [X] I agree to follow Biome's Code of Conduct

**Comment by ematipico** [comment]
We rarely add "special cases" in our rules, and we usually tend to add options instead.

**Comment by mrazauskas** [comment]
Got it. I agree with this policy.

My thinking was that in this context `vscode` is the host, not a library. Hence it might be worth treating the imports from `vscode` similarly how imports of `node:*` APIs are treated. In a way it sounds more like a pattern, than an exception.

Just wanted to leave few more thoughts. These are your internal decisions, of course. Definitely worth a discussion.

**Comment by bijela-gora** [comment]
@mrazauskas hi,

1. The package is actually published in npm: https://www.npmjs.com/package/vscode Maybe this can help you to resolve the issue.

2. In the playground the error is different and correct - no package.json

**Comment by mrazauskas** [comment]
@bijela-gora 

1. The link you provided is pointing to an ancient version of the package. It was last published five years ago and also is deprecated. Please double check the content of the links you are sharing.
2. If you know how to add `package.json` in the playground, please provided updated link. That would be useful indeed.

**Comment by bijela-gora** [comment]
@mrazauskas My purpose was not to solve your issue, but to show that the following statement isn't accurate.

> This is just a place holder, such package is not even published to npm registry.

Instead of using the playground it's also possible to prepare minimal git repo with reproduction like I did in this issue: https://github.com/biomejs/biome/issues/6733
If you prepare such git repo, I promise to check how the outdated vscode package can be used to solve your issue.

**Comment by mrazauskas** [comment]
Your suggestions were misleading. I reply only to warn others interested in this issue. 

Please stay constructive and do not waste my time. I did not ask for your assistance and I will not create any reproduction unless maintainers would ask for that.

**Comment by mrazauskas** [comment]
Actually the project where I have hit this problem does not have dependencies any more. I can disable the rule and close thus issue.

**Comment by ematipico** [comment]
I wonder if we should use `engines` to fix the issue. Here's how we do it in our VSCode extension, and I bet we would see the same issue, in case we enabled the rule

https://github.com/biomejs/biome-vscode/blob/main/package.json#L151-L153

**Comment by mrazauskas** [comment]
@ematipico Not sure about the `engines` field.

I think, `engines` is out of scope, because the description of the rule says:

> Disallow the use of dependencies that aren’t specified in the `package.json`.

---

If I create an empty repo, add Biome, run its `init` command,  and enable `noUndeclaredDependencies` run, creating a file with the following line does not raise any errors:

```js
import path from "node:path";

// ...
```

, but this errors:

```js
import * as vscode from "vscode";

// ...
```

I think, both should be treated the same. The reason is that neither `node:path`, nor `vscode` are not npm packages. There is no sense to check if they are specified in the `package.json` or not. (Actually, that would be a mistake, right?)

In other words, I think the `noUndeclaredDependencies` rule should only apply for the dependencies that are published to the npm registry. The rest is out of scope.

**Comment by bijela-gora** [comment]
> I think, both should be treated the same. 

👍
